predicators/envs/spot_env.py

```diff
@@ -671,7 +752,48 @@ def _get_next_nonpercept_atoms(self, obs: _SpotObservation,
         }
 
     def simulate(self, state: State, action: Action) -> State:
-        raise NotImplementedError("Simulate not implemented for SpotEnv.")
+        assert isinstance(action.extra_info, (list, tuple))
+        action_name, action_objs, _, _, action_fn, action_fn_args = \
+            action.extra_info
+        # The extra info is (action name, objects, function, function args).
+        # The action name is either an operator name (for use with nonpercept
+        # predicates) or a special name. See below for the special names.
+        assert self.action_space.contains(action.arr)
+
+        # Update the poses of the robot and all relevant objects
+        # to be in the correct position(s).
+        update_pbrspot_given_state(self.sim_robot, _obj_name_to_sim_obj, state)
+
+        # Special case: the action is "done", indicating that the robot
+        # believes it has finished the task. Used for goal checking.
+        if action_name == "done":
+            # During planning, trust that the goal is accomplished if the
+            # done action is returned.
+            self._current_task_goal_reached = True
+            return state
+
+        # Execute the action in the pybullet env. Automatically retry
+        # if a retryable error is encountered.
+        action_fn(*action_fn_args)  # type: ignore
```

Comment:
wait, now I'm very confused. what's the difference between the action function and the simulation function instead the action extra info?

Reply:
ohhh I see, this is the simulation function. change the name!